src/atlas/mesh/MeshBuilder.cc

```diff
@@ -16,6 +21,99 @@
+}
+
+// Check if grid points match mesh points -- not obviously true as the MeshBuilder sets the Grid
+// independently from the Mesh. This check can give confidence that the grid and mesh are specified
+// consistently with each other.
+//
+// This check makes a few assumptions
+// - the global_indices passed to the MeshBuilder are a 1-based contiguous index
+// - the Grid is one of StructuredGrid or UnstructedGrid. This restriction is because a
+//   CubedSphereGrid doesn't present a simple interface for the coordinates at the N'th point.
+// - the Mesh grid points are the grid nodes, and not the grid cell-centers (e.g., HEALPix or CubedSphere meshes)
+void validate_grid_vs_mesh(const atlas::Grid& grid, size_t nb_nodes, const double lons[], const double lats[],
+                           const int ghosts[], const gidx_t global_indices[], const eckit::mpi::Comm& comm) {
+    // Check assumption that global_indices look like a 1-based contiguous index
+    const size_t nb_owned_nodes = std::count(ghosts, ghosts + nb_nodes, 0);
+    size_t nb_nodes_global = 0;
+    comm.allReduce(nb_owned_nodes, nb_nodes_global, eckit::mpi::Operation::SUM);
+    for (size_t i = 0; i < nb_nodes; ++i) {
+        if (ghosts[i] == 0) {
+            // Check global_indices is consistent with a 1-based contiguous index over nodes
+            ATLAS_ASSERT(global_indices[i] >= 1);
+            ATLAS_ASSERT(global_indices[i] <= nb_nodes_global);
+        }
+    }
+
+    double lonlat[2];
+    const auto equal_within_roundoff = [](const double a, const double b) -> bool {
+        return fabs(a - b) <= 360.0 * 1.0e-16;
+    };
+
+    // Check lonlats for each supported grid type
+    if (grid.type() == "unstructured") {
+        const atlas::UnstructuredGrid ugrid(grid);
+        for (size_t i = 0; i < nb_nodes; ++i) {
+            if (ghosts[i] == 0) {
+                ugrid.lonlat(global_indices[i] - 1, lonlat);
+                if (!equal_within_roundoff(lonlat[0], lons[i]) || !equal_within_roundoff(lonlat[1], lats[i])) {
+                    throw_Exception("In MeshBuilder: UnstructuredGrid from config does not match mesh coordinates", Here());
+                }
+            }
+        }
+    } else if (grid.type() == "structured") {
+        const atlas::StructuredGrid sgrid(grid);
+        for (size_t i = 0; i < nb_nodes; ++i) {
+            if (ghosts[i] == 0) {
+                idx_t i, j;
+                sgrid.index2ij(global_indices[i] - 1, i, j);
+                sgrid.lonlat(i, j, lonlat);
+                if (!equal_within_roundoff(lonlat[0], lons[i]) || !equal_within_roundoff(lonlat[1], lats[i])) {
+                    throw_Exception("In MeshBuilder: StructuredGrid from config does not match mesh coordinates", Here());
+                }
+            }
+        }
+    } else {
+        throw_Exception("In MeshBuilder: can't validate a grid of type " + grid.type() + " from config");
+    }
```

Comment:
We can in principle use the iterator here with random access as last resort. It will work for any grid, but could be slower than the 2 above methods.
```c++
auto point = *(grid.lonlat().begin()+global_indices[i]-1);

Reply:
Good suggestion, I've added this